From 5db5b015d1474d7a1061dfd3d55a2199654bd3a3 Mon Sep 17 00:00:00 2001
From: Zhao Yihan <zhaoyihan@wxiat.com>
Date: Tue, 7 Jun 2022 17:42:45 +0800
Subject: [PATCH] sw64: remap PA with |= in early_ioremap
Patch-mainline: NO, OTHERS
References: openEuler-22.03-LTS
Modified-by-SEL: No


Sunway inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I56OLG

--------------------------------

BIOS passes logical address as physical address, which starts with
PAGE_OFFSET. It will overflow if add PAGE_OFFSET again.

Signed-off-by: Zhao Yihan <zhaoyihan@wxiat.com>

Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Acked-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/include/asm/early_ioremap.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/sw_64/include/asm/early_ioremap.h b/arch/sw_64/include/asm/early_ioremap.h
index 6f6fc6218cb3..930c6bf36ad3 100644
--- a/arch/sw_64/include/asm/early_ioremap.h
+++ b/arch/sw_64/include/asm/early_ioremap.h
@@ -14,7 +14,7 @@ early_ioremap(unsigned long phys_addr, unsigned long size)
 		y = (unsigned long) phys_to_virt(__pa(phys_addr));
 	} else {
 		y = phys_addr;
-		y += PAGE_OFFSET;
+		y |= PAGE_OFFSET;
 	}
 
 	return  (void __iomem *) y;
-- 
2.34.1

